### DISCUSSION TITLE
Extension of Value class

### DISCUSSION THREAD

**Discussion by malliaridis** [arkivanov/Decompose#389] [Ideas]
## Background

With the usage of `Value` it is possible to avoid direct dependency on coroutines / Reaktive. However, `Value` comes with some limitations since the implementation is not as extensive as coroutines or Reaktive.

One of the limitations I have faced so far is the **combination of multiple `Value`s** in order to get a single observable. This leads either to a custom function that replicates the basic idea of coroutine's `combine` or Reaktive's `combineLatest` function, or introduce direct dependency to coroutines or Reaktive by replacing `Value` with `Flow` or `Observable`.

Another extension function I had to add to my projects is the conversion of `Store.states` to `Value` in order to consume state updates from MVIKotlin in Decompose. This might be a very "specific" library use case, but then I started wondering if `Value` has to be part of Decompose and is not just a Decompose extension library or internal class.

With that said, I think that it is not necessary to remove `Value` or limit its scope, but a few more operations might come in handy.

## Discussion Topic

**I would like to discuss and maybe collect additional ideas for extending the `Value` class if that makes sense.**

One extension The extension I had in mind so far is a `combine` function that allows merging multiple `Value` instance into one. Even this extension might be not necessary if there are dependency-specific extensions (like converting `Flow` to `Value` or `Observable` to `Value`) where the merging can happen prior to the conversion to `Value` with the library-specific implementation.

**Comment by arkivanov** [comment]
Thanks for raising! Unfortunately, adding more operators for `Value` is very undesired. Because it's definitely not good for Decompose to open doors to become yet another reactive library. Also because `Value` allows reading the current value at any time and it is now thread safe, adding operators like `combine` would be pretty difficult.

It is advised (and you also mentioned this) to use `Flow` or `Observable`, perform all necessary transformations, and lastly convert to `Value`. So yeah, we can think about adding two more modules with coroutines and Reaktive extensions respectively.

As of extension for MVIKotlin `Store`, ideally this should be a separate library, but I can't afford one more. I would support if someone wants to create such a library, and we could even add it to MVIKotlin docs.

**Reply by malliaridis** [reply]
Thanks for the response and clarification. I think its important to be aware of this and decide early in a project which approach to follow. 🙂
